server/rest/src/main/java/org/infinispan/rest/resources/CacheResource.java

```diff
@@ -77,43 +80,36 @@ public Invocations getInvocations() {
             .create();
    }
 
-   private NettyRestResponse getCacheKeys(RestRequest request) throws RestResponseException {
-      try {
-         String cacheName = request.variables().get("cacheName");
+   private CompletionStage<RestResponse> getCacheKeys(RestRequest request) throws RestResponseException {
+      String cacheName = request.variables().get("cacheName");
 
-         String accept = request.getAcceptHeader();
-         if (accept == null) accept = MediaType.MATCH_ALL_TYPE;
+      String accept = request.getAcceptHeader();
+      if (accept == null) accept = MediaType.MATCH_ALL_TYPE;
 
-         MediaType contentType = negotiateMediaType(accept, cacheName);
-         AdvancedCache<Object, Object> cache = restCacheManager.getCache(cacheName, TEXT_PLAIN, TEXT_PLAIN);
-         CacheSet<Object> keys = cache.keySet();
-         Charset charset = Charset.fromMediaType(accept);
-         if (charset == null) charset = Charset.UTF8;
+      MediaType contentType = negotiateMediaType(accept, cacheName);
+      AdvancedCache<Object, Object> cache = restCacheManager.getCache(cacheName, TEXT_PLAIN, TEXT_PLAIN);
 
+      Charset mediaCharset = Charset.fromMediaType(accept);
+      Charset charset = mediaCharset == null ? Charset.UTF8 : mediaCharset;
+      return CompletableFuture.supplyAsync(() -> {
          NettyRestResponse.Builder responseBuilder = new NettyRestResponse.Builder();
          responseBuilder.contentType(contentType);
          responseBuilder.header(HttpHeaderNames.CACHE_CONTROL.toString(), CacheControl.noCache());
          OutputPrinter outputPrinter = EntrySetFormatter.forMediaType(contentType);
-         responseBuilder.entity(outputPrinter.print(cacheName, keys, charset));
+         responseBuilder.entity(outputPrinter.print(cacheName, cache.keySet(), charset));
```

Comment:
Sidetracking a little bit, I am also not a big fan of this method that returns all the keys to the user at once. It's here for historical reasons, plus it's also dangerous since it's associated directly with GET /rest/cacheName.

**Proposal**: remove this operation and put a note in the migration guide.

Reply:
I see chunking is gone in HTTP/2, so this is definitely worth a separate issue :)

Reply:
![Http/3](https://cdn-images-1.medium.com/max/800/1*z5o-mslkNcdDbGzo6G3zqg.jpeg)